mock/exchange.go

```diff
@@ -15,16 +23,63 @@ func NewExchange(seed float64) *Exchange {
 
 // Sets prices for upcoming orders.
 func (m *Exchange) SetPrices(prices trading.Window) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 	m.prices = prices
+
+	m.applyOrders()
 }
 
 // Returns current portfolio.
 func (m *Exchange) CurrentPortfolio() trading.Portfolio {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 	return m.portfolio
 }
 
 // Simulates an order based on a trading window during which it was placed. Must call SetPrices
 // first. Updates portfolio accordingly.
 func (m *Exchange) SubmitOrder(order trading.Order) trading.OrderOutcome {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	switch order.Type {
+	case "limit", "stop":
+		m.orders = append(m.orders, order)
+	default:
+		panic(fmt.Errorf("uknown order type %s", order.Type))
+	}
+
+	m.applyOrders()
+
 	return trading.OrderOutcome{}
 }
+
+func (m *Exchange) applyOrders() {
+	unfilledOrders := []trading.Order{}
+	for _, order := range m.orders {
+		ticker := order.Ticker
+		price := m.prices.Close[ticker]
+		orderCost := price * float64(order.Volume)
+
+		switch order.Type {
+		case "limit":
+			if price >= order.Price && orderCost <= m.portfolio.CashUSD {
+				m.portfolio.Stocks[ticker] += order.Volume
+				m.portfolio.CashUSD -= orderCost
+			} else {
+				unfilledOrders = append(unfilledOrders, order)
+			}
+		case "stop":
+			if price <= order.Price {
```

Comment:
I'm ignorant about the types of orders we could have. But should we have both stop and limit for buy and sell orders?

Also, do I have this backwards? I would think if the price is higher than order.Price that is more favorable condition for selling. Like I want to protect against selling the stock for $0.

Reply:
I'll look into the order types a bit more.

Looks like there's `Buy/Sell` variations of each of the order types.

https://www.thebalance.com/making-sense-of-day-trading-order-types-1031387

Should we implemented all of these or just do a simpler buy market / sell market for the first pass and add all the other types later? I'm open to either.

Reply:
We can start small.

I think I was thinking of a stop-limit order, where we would have a stop like this and a limit below the stop price where we would not sell below the limit (e.g., protect against selling at extremely low prices).

We can start with these, but we should at least add `Buy/Sell` to `trading.Order`?

Reply:
I can add all the order types. Shouldn't be too difficult. I'll do that now.

Reply:
Added a few more things here. It's becoming even more clear to me that we need a more sophisticated simulator.